### PR TITLE
Add "default_value" parameter

### LINKED ISSUE #1324: Add a way to generate `Default` for enums
It would be great to annotate the default value in `Gir.toml` and have gir generate the corresponding `impl Default`.

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
Fixes #1324.

For now this is a draft because:

 * Is the parameter name ok? If not what should we use?
 * I chose to let the user put whatever they wanted as default value. Did you have something else in mind?
 * I need to add documentation.

### CODE REVIEW

File: src/codegen/enums.rs [hunk 1]
```diff
@@ -6,6 +6,7 @@ use crate::{
         self, cfg_condition, cfg_condition_no_doc, cfg_condition_string, cfg_deprecated, derives,
         doc_alias, version_condition, version_condition_no_doc, version_condition_string,
     },
+    codegen::object::generate_default_impl,
```
Comment (sdroege) [inline src/codegen/enums.rs:9]: Why do you move it to the `object` module? Has nothing to do with objects :)

File: src/config/gobjects.rs [hunk 2]
```diff
@@ -417,6 +420,10 @@ fn parse_object(
         .lookup("clear_function_expression")
         .and_then(Value::as_str)
         .map(ToOwned::to_owned);
+    let default_impl_value = toml_object
+        .lookup("default_impl_value")
```
Comment (sdroege) [inline src/config/gobjects.rs:424]:
```suggestion
        .lookup("default_value")
```

Seems more descriptive

**Review by sdroege — COMMENTED** [review]
Looks good to me

File: src/config/gobjects.rs (hunk 2 above)
Comment (bilelmoussaoui) [reply src/config/gobjects.rs:424]: default_value looks good to me as well

File: src/codegen/enums.rs (hunk 1 above)
Comment (GuillaumeGomez) [reply src/codegen/enums.rs:9]: Where should I put it then?

Comment (sdroege) [reply src/codegen/enums.rs:9]: `codegen/mod.rs`? And maybe call it `generate_enum_flags_default_impl`

Comment (GuillaumeGomez) [reply src/codegen/enums.rs:9]: Oh you only want it for enums and flags? Not for structs?

Comment (sdroege) [reply src/codegen/enums.rs:9]: How would this work properly for structs?

Comment (GuillaumeGomez) [reply src/codegen/enums.rs:9]: Exactly the same. It's expecting an expression to put into the `default` function.

Comment (bilelmoussaoui) [reply src/codegen/enums.rs:9]: Ah that is not really what I was thinking of, it should take a member name and check if it exists or error out. Then return `Self::MemberName` in the default impl.

Comment (GuillaumeGomez) [reply src/codegen/enums.rs:9]: Hum ok. I'll send an error if objects have this attribute then.

**Review by sdroege — APPROVED** [review]

File: src/codegen/mod.rs [hunk 3]
```diff
@@ -97,3 +100,40 @@ pub fn generate_single_version_file(env: &Env) {
...
+
+pub fn generate_default_impl<'a>(
+    w: &mut dyn Write,
+    env: &Env,
+    config: &crate::config::gobjects::GObject,
+    name: &str,
+    version: Option<crate::version::Version>,
+    mut members: impl Iterator<Item = &'a str>,
+) -> Result<()> {
+    if let Some(ref default_value) = config.default_value {
```
[the remark this answers is not in the document]
Comment (bilelmoussaoui) [reply src/codegen/mod.rs:112]: I would rather be consistent with the other configurations we have :)

[the remark this answers is not in the document]
Comment (GuillaumeGomez) [reply src/codegen/mod.rs:112]: So for `gtk::Align::Fill` it'd be the C name `GTK_ALIGN_FILL`? It's a const after all...

[the remark this answers is not in the document]
Comment (bilelmoussaoui) [reply src/codegen/mod.rs:112]: the name is `fill`

[the remark this answers is not in the document]
Comment (GuillaumeGomez) [reply src/codegen/mod.rs:112]: @sdroege An opinion maybe?

[the remark this answers is not in the document]
Comment (bilelmoussaoui) [reply src/codegen/mod.rs:112]: It confused me as well the first time i had to set a specific version of an enum member but at least it is consistent

[the remark this answers is not in the document]
Comment (sdroege) [reply src/codegen/mod.rs:112]: What @bilelmoussaoui said. That would be consistent with how other configurations for enums/flags/signals/etc work.

**Review by bilelmoussaoui — APPROVED** [review]

**Pull request merged**